hack/generator/spec/versioning.md

```diff
@@ -0,0 +1,201 @@
+### Use the latest available API Version
+
+As a starting point for defining the storage schema for each resource type, we will use the latest available version of the matching ARM API. This ensures maximum compatibility with that API when it is used, giving a good starting point for new adoption of the service operator.
+
+If a resource type has been dropped from the ARM API, we will still generate a storage schema for it based on the last ARM API version where it existed; this ensures backward compatibility with existing service operator deployments.
+
+### Augment storage with a name/value property bag
+
+To facilitate easy and safe storage of additional configuration in storage, we'll include a name/value property bag for storage, along with helper methods that allow for strongly typed access to values stored within.
+
+Sequestering the information away within the storage schema is more robust than using separate annotations as they are less subject to arbitary modification by users. This allows us to largely avoid situations where well meaning (but uninformed) consumers of the service operator innocently make changes that result in the operator becoming failing. We particularly want to avoid this failure mode because recovery will be difficult - restoration of the modified/deleted information may be impractical or impossible.
+
+---
+
+****Outstanding issue:** Where do the helper functions for the name/value property bag reside?**
+
+We could generate those anew for each type, making them conveniently available and avoiding the need for our generated types to depend on anything more than the standard Go library Alternatively, we could define the property bag as a new struct, wrapping up storage & helper methods into a reusable form.
+
+---
+
+### Generate conversion methods
+
+Each of the structs generated for ARM API will have the normal `ConvertTo()` and `ConvertFrom()` methods generated automatically, mapping like-for-like properties between the hub version and the current type.
+
+* When the name and type of properties match exactly, conversion in both directions will be simple assignments.
+* Where the name matches but the type does not, the property will be skipped by default.
+* If missing on the hub version, the name/value property bag will also be used.
+
+### Configuration for common changes
+
+We'll facilitate common types of changes between versions by providing explicit configuration options.
+
+**Property rename** - if a property is simply renamed between versions (say for a spelling correction), we can generate the required assignment for compatiblity between spoke and hub versions.
+
+**Type rename** - if a subtype is renamed between versions, maintaining largely the same structure, we can automatically map between those two types.
+
+
+---
+
+***Outstanding Issue:** What do we do for other kinds of version-to-version change?*
+
+Are there other cases of changes between versions that we may be able to handle automatically. 
```

Comment:
Well, v3 is still the storage verison, right? As that's the one you're storing... To convert from v2 to v1, it would be 
v2 -> v3 -> v1. The conversion from v2 -> v3 needs to handling converting the real type to the property bag representation and then v3 -> v1 needs to do the opposite. Or am I missing something?

Reply:
No you have it right Dave -- my point is just that when we put the `v2` representation into the property bag, it's still a `v2` type, not a `v3` type (since there is no `v3` version of that type). So as far as _that particular type_ is concerned, it's kinda still the storage version and all conversions for that type need to go through `v2` (not `v3`, because there is no corresponding type in `v3`).

Basically I am saying I think we're going to end up in a world where: the storage version is not per-api (i.e. what we in Azure think of as an API - a related set of resources) rather it's per-type. This is basically what @porges was saying in the call Bevan/I with him earlier I think

Reply:
Not sure I'm fully understanding the challenge. Might be good to discuss in this week's call.

Reply:
When a type is removed, the latest version of that type will be used to define the storage version. Just as we do for other types, we convert to the storage type for storage - we'll put it into the PropertyBag as an `interface{}` and rely on the way it's serialized by the bag.